### PR TITLE
iox-#765 list stl dependencies with bash script

### PR DESCRIPTION
**Pull request by elfenpiff**
Signed-off-by: Christian Eltzschig <me@elchris.org>

## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
Introducing the script `tools/list_stl_dependencies.sh` which lists all STL dependencies with 3 sections.
1. What STL constructs are we using
2. Where do we set `using namespace std`
3. Which file uses which STL construct

You can test the script with
```sh
cd ICEORYX_ROOT_DIR
./tools/list_stl_dependencies.sh
```

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- #765

### CODE REVIEW

File: tools/list_stl_dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+
+
+#=====================================================
+# usage:
+#   cd ICEORYX_ROOT_DIR
+#   ./tools/list_stl_dependencies.sh
+#=====================================================
+
```
Comment (dkroenke) [inline tools/list_stl_dependencies.sh:25]: You could add something like `WORKSPACE=$(git rev-parse --show-toplevel)` to step automatically into the iceoryx base folder.

File: tools/list_stl_dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+SOURCE_DIR=(source include)
+
+
+for COMPONENT in ${COMPONENTS[@]}; do
+    for DIR in ${SOURCE_DIR[@]}; do
+        GREP_PATH="${GREP_PATH} ${COMPONENT}/$DIR"
+    done
+done
+
+echo
```
Comment (dkroenke) [inline tools/list_stl_dependencies.sh:36]: maybe the output can be piped additionally into a file.

File: tools/list_stl_dependencies.sh (hunk 1 above)
Comment (elfenpiff) [reply tools/list_stl_dependencies.sh:24]: Nice idea ... it is done!

File: tools/list_stl_dependencies.sh (hunk 2 above)
Comment (elfenpiff) [reply tools/list_stl_dependencies.sh:35]: Yes this is possible with the normal bash tooling. Just write
```sh
./list_stl_dependencies.sh > someFile.txt
```

**Review by dkroenke — APPROVED** [review]

File: tools/list_stl_dependencies.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+echo files with stl dependency
+grep -RIne "std::" $GREP_PATH | sed -n  "s/\([^:]*\)\:[0-9]*\:.*std::[a-zA-Z_]*.*/\1/p" | xargs -I{} basename {} | sort | uniq
+
+echo
+echo using namespace with std component
+grep -RIne ".*using[ ]*namespace[ ]*std" $GREP_PATH | sed -n "s/\(.*\)/\ \ \1/p"
+
+echo
+echo usage of std components
+grep -RIne "std::" $GREP_PATH | sed -n  "s/.*\(std::[a-zA-Z_]*\).*/\ \ \1/p" | sort | uniq
```
Comment (mossmaurice) [inline tools/list_stl_dependencies.sh:49]: @dkroenke We could add a whitelist of allowed STL components here and run in CI.

**Review by mossmaurice — APPROVED** [review]
LGTM

**Pull request merged**